### DISCUSSION TITLE
Android VPN Client

### DISCUSSION THREAD

**Discussion by Smart781** [OutlineFoundation/outline-sdk#267] [Q&A]
Hello! I am writing an Outline VPN client for Android in the Go language. I managed to write a client for desktop systems, but on Android I had the following problem. I use outline-go-tun2SOCKS (https://github.com/Jigsaw-Code/outline-apps/tree/master/client/go). In order to connect with the server on Android, I need a TUN file descriptor (https://github.com/Jigsaw-Code/outline-apps/blob/master/client/go/outline/tun2socks/tunnel_android.go#L43) is created on Android using VpnService (https://developer.android.com/reference/android/net/VpnService). But it cannot be launched from Go (https://pkg.go.dev/github.com/eycorsican/go-tun2socks-android), so I try to use Gomobile Bind (https://pkg.go.dev/golang.org/x/mobile/cmd/gobind) to create a method working on Go and calling vpnservice, which is made
on Kotlin. However, during the compilation of Gomobile Bind, errors occur. I asked this question on another forum and was suggested to use RunNative to access OS level APIs, but I will be glad to any solutions. Could you please, help me with that? Any help would be much appreciated!!

**Comment by jyyi1** [comment]
Hi @Smart781 , 

I think the [go-tun2socks-android](https://pkg.go.dev/github.com/eycorsican/go-tun2socks-android) is not maintained any more.

Here are two potential solutions you might want to consider:

- You could try [gomobile's reverse binding](https://pkg.go.dev/golang.org/x/mobile/cmd/gobind#hdr-Reverse_bindings) to call `VpnService` from Go.
- If reverse binding doesn't work, you may also try exporting `VpnService` through C using the Java Native Interface (JNI). Then, you can leverage cgo in Go to call these exported C functions.
